src/StatCache.cc

```diff
@@ -0,0 +1,65 @@
+/*
+ *  OpenSCAD (www.openscad.org)
+ *  Copyright (C) 2009-2011 Clifford Wolf <clifford@clifford.at> and
+ *                          Marius Kintel <marius@kintel.net>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  As a special exception, you have permission to link this program
+ *  with the CGAL library and distribute executables, as long as you
+ *  follow the requirements of the GNU GPL in regard to all of the
+ *  software in the executable aside from CGAL.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include "StatCache.h"
+#include "printutils.h"
+
+#include <sys/stat.h>
+#include <string>
+#include <unordered_map>
+
+typedef std::unordered_map<std::string, struct stat> StatMap;
+
+static StatMap statMap;

```

Comment:
I'm not concerned with optimization opportunities by sharing the cache between windows.
My concern was that the automatic reload for two windows using an overlapping file set could run at approximately the same time, causing both to use the cache during the same traversal, thus invalidating the cache at unfortunate points in time. This could happen since the dependency checking is driven by timers, and the event loop may be processed during a checking operations (by calling `QCoreApplication::processEvents()` causing more than one MainWindow to process dependencies simultaneously).

Looking closely at the code, I _think_ this should never happen, but the logic for managing this is very convoluted (through flags for blocking `processEvents()` as well as global GUI lock - could really benefit from a cleanup).
..so what you've written should work for now, but would need a refactoring at the same time as the mess mentioned above.